tts: add speaker file support

dm4 opened the pull request:
- Added support for TTS speaker files, including a new command-line option `--tts-speaker-file` to specify the file path.
- Implemented JSON handling in `tts.cpp` to load and parse speaker data, enhancing audio generation capabilities.

dm4 committed:
tts: add speaker file support

Signed-off-by: dm4 <sunrisedm4@gmail.com>

ngxson commented on common/arg.cpp:
```diff
@@ -2447,6 +2447,13 @@ common_params_context common_params_parser_init(common_params & params, llama_ex
+        [](common_params & params, const std::string & value) {
+            params.vocoder.speaker_file = value;
+        }
+    ).set_examples({LLAMA_EXAMPLE_TTS, LLAMA_EXAMPLE_SERVER}));
```
```suggestion
    ).set_examples({LLAMA_EXAMPLE_TTS}));
```

I don't think llama-server supports this for now

dm4 replied:
Thank you for your suggestion. I have removed `LLAMA_EXAMPLE_SERVER`.

Koalamana9 commented:
For those who are interested, the speaker files: https://github.com/edwko/OuteTTS/tree/main/outetts/version/v1/default_speakers
For any non english text make sure to romanize prompt

ngxson commented:
@edwko Could you please have a look on this PR?

dm4 committed:
tts: handle outetts-0.3

dm4 commented:
Hello @ngxson and @edwko, I have already added support for version 0.3. Since `common_get_builtin_chat_template()` was removed in [this commit](https://github.com/ggml-org/llama.cpp/commit/6171c9d25820ccf676b243c172868819d882848f), I have switched to using `llama_model_chat_template()` to obtain the model's `tokenizer.chat_template` metadata.

ngxson commented on examples/tts/tts.cpp:
```diff
@@ -452,6 +456,75 @@ static std::vector<llama_token> prepare_guide_tokens(const llama_vocab * vocab,
+    return speaker;
+}
+
+static std::string get_tts_version(llama_model *model, json speaker = json::object()) {
```
small nits here, maybe better if we use an enum for version number:

```
enum outetts_version {
    OUTETTS_V0_2,
    OUTETTS_V0_3,
};
```

dm4 replied:
Thanks for the reminder! This part has been fixed.

Koalamana9 commented:
@ggerganov merge it please

ggerganov commented:
Can you provide examples commands both for v0.2 and v0.3 so I can run some tests?

Koalamana9 commented:
Example commands for v0.2 and v0.3 are identical: -m OuteTTS-v2-or-v3 -mv Wavtokenizer -c 4096 --tts-use-guide-tokens --tts-speaker-file en_female_1.json -p "Hello world"
Speaker files from here: https://github.com/edwko/OuteTTS/tree/main/outetts/version/v1/default_speakers
OuteTTS v0.3: https://huggingface.co/OuteAI/OuteTTS-0.3-500M-GGUF/tree/main
OuteTTS v0.2: https://huggingface.co/OuteAI/OuteTTS-0.2-500M-GGUF/tree/main
Wavtokenizer: https://huggingface.co/novateur/WavTokenizer-large-speech-75token/tree/main (must be converted to gguf)

--tts-use-guide-tokens is optional, sometimes gives better results for v0.2

For prompts longer than 10 words it can hit this assert and stop generation
https://github.com/ggml-org/llama.cpp/blob/14dec0c2f29ae56917907dbf2eed6b19438d0a0e/src/llama.cpp#L8470
Removing this assert allows for longer prompt generation.

ggerganov reviewed (approved):
The assert triggers because the audio codes exceed the size of the microbatch (default 512). The vocoder uses non-causal attention, so it requires that all codes fit in a single microbatch. The workaround is to add `-ub 4096` to your arguments and the proper solution is to create the WavTokenizer's context with `n_ubatch` equal to the `n_ctx_train()` of the model.

ggerganov committed:
tts : add new line in error message

Koalamana9 commented:
Awesome! With OuteTTS v0.3 it even generates all punctuation correctly! To be honest this is already quite a good quality for such a small model. Perhaps it is worth updating examples/tts/README and adding -ub 4096 argument as it is necessary for correct generation. I would like to see more PR's merged in this, for example #11070 server example really works well without unloading models from memory. It is possible to further develop this example as it is of great value for many, especially considering that OuteTTS 0.3-500M is allowed for commercial use.
Despite the fact that Kokoro TTS is now considered the highest quality text to speech model with the smallest size, it is worth noting that it's a very distilled model without the ability to emulate voices except for simply blending existing ones, meanwhile OuteTTS offers a novel approach to speech synthesis using simple llm where anyone can generate a copy of a voice in just a few seconds by passing simple json file.

ggerganov commented:
Feel free to contribute improvements. I think the `tts` example is in a very hacky state and can be improved in many ways. Ideally, it should become a more general purpose TTS example that would support more TTS models. But we first need the infra for that to be added to `libllama`, which I am working on atm.

Also, I think figuring out streaming first is crucial before making major changes and additions.
